### PR TITLE
Handle segfaults in the profiler on Windows

### LINKED ISSUE #10638: Profiler segfaults on windows
This code makes the profiler segfault on windows
```
function test_func()
    points_1= Vector{Float64}[[0.1, 0.,-0.4],[-0.1, 0., -0.4 ], 	
                 [0.1, 0.325, 0.117],  [-0.1,0.325, 0.117 ],		
                 [ 0.1, 7., -0.9], [ -0.1, 7., -0.9]] 
    rot = [cos(1) -sin(1) 0
           sin(1) cos(1) 0
           0 0 1]
    line_crash = [(rot * vec) + [0,1.97,0] for vec in points_1] #If this lines is commented out it don't segfault
end

function test(n)
    tic()
    for i=1:n
        test_func()
    end
    toc()
end
test(10)
Profile.clear()
@profile test(100000) # If this number is reduced then it stops segfaulting
Profile.print()
```
The output is:
```
julia> @profile test(100000) # If this number is reduced then it stops segfaulting

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x6c525aee -- rec_backtrace_ctx at C:\Users\Hoegh\Julia-0.4.0-dev_26_03\bin\libjulia-debu
g.dll (unknown line)
rec_backtrace_ctx at C:\Users\Hoegh\Julia-0.4.0-dev_26_03\bin\libjulia-debug.dll (unknown line)
jl_typeof at C:\Users\Hoegh\Julia-0.4.0-dev_26_03\bin\libjulia-debug.dll (unknown line)
BaseThreadInitThunk at C:\Windows\system32\kernel32.dll (unknown line)
RtlUserThreadStart at C:\Windows\SYSTEM32\ntdll.dll (unknown line)
```
My version info is:
```
C:\Users\Hoegh>Julia-0.4.0-dev_26_03\bin\julia-debug.exe
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+3981 (2015-03-25 02:18 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 237b811* (1 day old master)
|__/                   |  x86_64-w64-mingw32

julia> versioninfo()
Julia Version 0.4.0-dev+3981
Commit 237b811* (2015-03-25 02:18 UTC)
DEBUG build
Platform Info:
  System: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-2630QM CPU @ 2.00GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3

julia>
```

### PR DESCRIPTION
**Pull request by ihnorton**
This one needs close review for obvious reasons @vtjnash @Keno 

Microsoft [says](https://msdn.microsoft.com/en-us/library/xdkz3x12.aspx) it is not legal to `longjmp` in a signal handler (except for SIGFPE)...

fixes #10638

### CODE REVIEW

File: src/profile.c [hunk 1]
```diff
@@ -44,22 +65,38 @@ static DWORD WINAPI profile_bt( LPVOID lparam )
                 fputs("failed to suspend main thread. aborting profiling.",stderr);
                 break;
             }
-            CONTEXT ctxThread;
+
+            LPTOP_LEVEL_EXCEPTION_FILTER prevHandler = SetUnhandledExceptionFilter(profile_exception_handler);
+
+            CONTEXT ctxThread __attribute__((aligned(4)));
```
Comment (ihnorton) [inline src/profile.c]: ah, whoops. didn't mean to include this. but, FYI: there is a bunch of verbiage in every discussion I read saying that `CONTEXT` must be DWORD-aligned (probably is already by default with gcc?)

**Comment by ihnorton** [comment]
After some more reading, I think most of the reasons it would be unsafe to `longjmp` here don't apply: there is no heap allocation here, there is no program state being touched, we reset the signal handler, etc.

**Comment by vtjnash** [comment]
if you change it to `jl_setjmp`/`jl_longjmp`, it would be legal.

we might be better off just checking if the pointer is valid via VirtualQuery in the first place:
https://msdn.microsoft.com/en-us/library/windows/desktop/aa366902(v=vs.85).aspx

or precompute the bounds of the stack (again by using VirtualQuery, but starting from some stack address (such as `jl_stackbase`)

**Comment by ihnorton** [comment]
> we might be better off just checking if the pointer is valid via VirtualQuery in the first place:

`Context->Rip` in rec_backtrace?

**Comment by vtjnash** [comment]
`Context->Rsp` (before we try to dereference it https://github.com/ihnorton/julia/blob/prof_segv_handler/src/task.c#L606)

**Comment by ihnorton** [comment]
Adding VirtualQuery check does resolve the issue. I left the segfault handler commits in too in case we still want that. Tested separately.

**Comment by vtjnash** [comment]
I'm a little nervous about changing the handler. We should at least reset it before resuming the thread however. Also, please add a comment on the lines you added

What is the performance penalty of this on backtrace collection:
f() = f()
@time try f() end

**Comment by ihnorton** [comment]
(after warmup)
* master:                                `elapsed time: 0.019003433 seconds (160 bytes allocated)`
* virtualquery check only:      `elapsed time: 0.018860939 seconds (160 bytes allocated)`
* segv handler only:              `elapsed time: 0.019733667 seconds (160 bytes allocated)`  -- (this was highest. other runs were .190)
* both (this pr):                      `elapsed time: 0.018887385 seconds (160 bytes allocated)`

> We should at least reset it before resuming the thread however. Also, please add a comment on the lines you added

done and done.

File: src/task.c [hunk 2]
```diff
@@ -600,9 +600,20 @@ DLLEXPORT size_t rec_backtrace_ctx(ptrint_t *data, size_t maxsize, CONTEXT *Cont
...
+
         PRUNTIME_FUNCTION FunctionEntry = (PRUNTIME_FUNCTION)JuliaFunctionTableAccess64(GetCurrentProcess(), Context->Rip);
         if (!FunctionEntry) { // assume this is a NO_FPO RBP-based function
             Context->Rsp = Context->Rbp;                 // MOV RSP, RBP
+
+            // Check whether the pointer is valid and executable before dereferencing
+            // to avoid segfault while recording. See #10638.
+            if (VirtualQuery((LPCVOID)Context->Rsp, &mInfo, sizeof(MEMORY_BASIC_INFORMATION)) == 0)
+                break;
+            if (mInfo.AllocationProtect < PAGE_EXECUTE)
```
Comment (vtjnash) [inline src/task.c]: > The access granted to all pages is the same (PAGE_READONLY, PAGE_READWRITE, PAGE_NOACCESS, PAGE_WRITECOPY, PAGE_EXECUTE, PAGE_EXECUTE_READ, PAGE_EXECUTE_READWRITE, PAGE_EXECUTE_WRITECOPY, PAGE_GUARD, or PAGE_NOCACHE).
https://msdn.microsoft.com/en-us/library/windows/desktop/aa366786(v=vs.85).aspx

I think this should be `X = mInfo.AllocationProtect; ((X&PAGE_READONLY) || (X&PAGE_READWRITE) || (X&PAGE_WRITECOPY) || (X&PAGE_EXECUTE_READ)) && !((X&PAGE_GUARD) || (X&PAGE_NOACCESS))`

Comment (ihnorton) [inline src/task.c]: ok, done (with logic fix)

**Comment by ihnorton** [comment]
I merged the VirtualQuery part since that was uncontroversial. We can revisit this later if we decide the segfault handler would be useful.

**Pull request closed without merging**